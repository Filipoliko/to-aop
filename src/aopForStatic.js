import createGetTrap from './trap/createGetTrap';
import createSetTrap from './trap/createSetTrap';
import overOwnProperty from './overOwnProperty';
import { AOP_STATIC_ALLOW, AOP_HOOKS } from './symbol';

export default function aopForStatic(target, pattern) {
  let original = {};
  if (!Object.prototype.hasOwnProperty.call(target, AOP_STATIC_ALLOW)) {
    Reflect.defineProperty(target, AOP_STATIC_ALLOW, {
      value: false,
      enumerable: false,
      writable: true
    });
  } else {
    target[AOP_STATIC_ALLOW] = false;
  }
  let originalTarget = target;

  while (target && target !== Function.prototype) {
    // TODO improve
    Object.entries(Object.getOwnPropertyDescriptors(target)).forEach(
      ([property, descriptor]) => {
        if (
          typeof descriptor.get === 'function' ||
          typeof descriptor.set === 'function'
        ) {
          if (!Object.prototype.hasOwnProperty.call(original, property)) {
            original[property] = target[property];
          }
          Reflect.defineProperty(
            target,
            property,
            Object.assign({}, descriptor, {
              get: (...rest) => {
                if (originalTarget[AOP_STATIC_ALLOW]) {
                  let aopHooks = descriptor.get[AOP_HOOKS];
                  if (aopHooks) {
                    const { object } = aopHooks[aopHooks.length - 1];

                    aopHooks.push({
                      target,
                      object,
                      property,
                      pattern
                    });

                    return descriptor.get(...rest);
                  }

                  return createGetTrap({
                    target,
                    object: original,
                    property,
                    pattern
                  })(...rest);
                }
              },
              set: payload => {
                if (originalTarget[AOP_STATIC_ALLOW]) {
                  return createSetTrap({
                    target,
                    object: original,
                    property,
                    pattern
                  })(payload);
                }
              }
            })
          );
        }

        // if (
        //   typeof target[property] === 'function' &&
        //   (!isConstructable(target[property]) || target[property][AOP_HOOKS])
        // ) {
        //   original[property] = target[property];
        //
        //   let aopHooks = original[property][AOP_HOOKS];
        //   if (aopHooks) {
        //     const { object } = aopHooks[aopHooks.length - 1];
        //     aopHooks.push({
        //       target,
        //       object,
        //       property,
        //       pattern
        //     });
        //     return;
        //   }
        //
        //   target[property] = createCallTrap({
        //     target,
        //     object: original,
        //     property,
        //     pattern
        //   });
        // }
      }
    );

    overOwnProperty({ target, pattern, original, object: target });
    target = Reflect.getPrototypeOf(target);
  }
  originalTarget[AOP_STATIC_ALLOW] = true;
}
