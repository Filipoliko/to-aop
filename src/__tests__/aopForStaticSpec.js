import aopForStatic from '../aopForStatic';

function createClasses() {
  let staticSetter = null;

  class A {
    static staticMethod() {
      return 'static method';
    }

    static get staticGetter() {
      return 'static getter';
    }

    static set staticSetter(value) {
      staticSetter = value;
    }

    static get staticSetter() {
      return staticSetter;
    }

    constructor(variable) {
      this.variable = variable;
    }

    method() {
      return this.variable;
    }
  }

  class B extends A {
    constructor(variable) {
      super(variable);
    }

    method(...rest) {
      return 'B ' + super.method(...rest);
    }
  }

  class C extends A {
    static staticMethod2() {
      return 'static method 2';
    }

    constructor(variable) {
      super(variable);

      this.map = new Map();
    }

    method2() {
      return this.map;
    }
  }

  return { A, B, C };
}

describe('aopForStatic method', () => {
  let pattern = null;
  let pattern2 = null;

  beforeEach(() => {
    pattern = {
      beforeMethod: () => {},
      afterMethod: () => {},
      beforeGetter: () => {},
      beforeSetter: () => {},
      afterGetter: () => {},
      afterSetter: () => {}
    };
    pattern2 = Object.assign({}, pattern);
  });

  describe('parent classs with static method', () => {
    let result = null;

    beforeEach(() => {
      spyOn(pattern, 'beforeMethod');
      spyOn(pattern2, 'beforeMethod');
      spyOn(pattern, 'afterMethod');
      spyOn(pattern2, 'afterMethod');

      const { A } = createClasses();

      aopForStatic(A, pattern);
      aopForStatic(A, pattern2);

      result = A.staticMethod();
    });

    it('should call pattern.beforeMethod and pattern.afterMethod', () => {
      expect(pattern.beforeMethod.calls.count()).toEqual(1);
      expect(pattern.beforeMethod.calls.argsFor(0)).toMatchSnapshot();
      expect(pattern.afterMethod.calls.count()).toEqual(1);
      expect(pattern.afterMethod.calls.argsFor(0)).toMatchSnapshot();
      expect(result).toEqual('static method');
    });

    it('should call pattern2.beforeMethod and pattern2.afterMethod', () => {
      expect(pattern2.beforeMethod.calls.count()).toEqual(1);
      expect(pattern2.beforeMethod.calls.argsFor(0)).toMatchSnapshot();
      expect(pattern2.afterMethod.calls.count()).toEqual(1);
      expect(pattern2.afterMethod.calls.argsFor(0)).toMatchSnapshot();
      expect(result).toEqual('static method');
    });
  });

  describe('parent classs with static getter', () => {
    let result = null;

    beforeEach(() => {
      const { A } = createClasses();

      aopForStatic(A, pattern);
      aopForStatic(A, pattern2);

      spyOn(pattern, 'beforeGetter');
      spyOn(pattern2, 'beforeGetter');
      spyOn(pattern, 'afterGetter');
      spyOn(pattern2, 'afterGetter');

      result = A.staticGetter;
    });

    it('should call pattern.beforeGetter and pattern.afterGetter', () => {
      expect(pattern.beforeGetter.calls.count()).toEqual(1);
      expect(pattern.beforeGetter.calls.argsFor(0)).toMatchSnapshot();
      expect(pattern.afterGetter.calls.count()).toEqual(1);
      expect(pattern.afterGetter.calls.argsFor(0)).toMatchSnapshot();
      expect(result).toEqual('static getter');
    });

    it('should call pattern2.beforeGetter and pattern2.afterGetter', () => {
      expect(pattern2.beforeGetter.calls.count()).toEqual(1);
      expect(pattern2.beforeGetter.calls.argsFor(0)).toMatchSnapshot();
      expect(pattern2.afterGetter.calls.count()).toEqual(1);
      expect(pattern2.afterGetter.calls.argsFor(0)).toMatchSnapshot();
      expect(result).toEqual('static getter');
    });
  });
});
