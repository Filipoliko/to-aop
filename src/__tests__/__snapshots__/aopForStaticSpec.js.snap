// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`aopForStatic method should call pattern.beforeGetter and pattern.afterGetter for inhereted static getter 2`] = `
Array [
  Object {
    "context": undefined,
    "object": Object {
      "staticMethod": [Function],
      "staticMethod2": [Function],
    },
    "property": "staticGetter",
    "target": [Function],
  },
]
`;

exports[`aopForStatic method should call pattern.beforeGetter and pattern.afterGetter for inhereted static getter 3`] = `
Array [
  Object {
    "context": undefined,
    "object": Object {
      "staticMethod": [Function],
      "staticMethod2": [Function],
    },
    "payload": "static getter",
    "property": "staticGetter",
    "target": [Function],
  },
]
`;

exports[`aopForStatic method should call pattern.beforeGetter and pattern.afterGetter for inhereted static getter with multiple aspect 2`] = `
Array [
  Object {
    "context": undefined,
    "object": Object {
      "staticMethod": [Function],
      "staticMethod2": [Function],
    },
    "property": "staticGetter",
    "target": [Function],
  },
]
`;

exports[`aopForStatic method should call pattern.beforeGetter and pattern.afterGetter for inhereted static getter with multiple aspect 3`] = `
Array [
  Object {
    "context": undefined,
    "object": Object {
      "staticMethod": [Function],
    },
    "payload": "static getter",
    "property": "staticGetter",
    "target": [Function],
  },
]
`;

exports[`aopForStatic method should call pattern.beforeMethod and pattern.afterMethod for inhereted static method with multiple aspect 1`] = `
Array [
  Object {
    "args": Array [],
    "context": [Function],
    "object": Object {
      "staticMethod": [Function],
    },
    "property": "staticMethod",
    "target": [Function],
  },
]
`;

exports[`aopForStatic method should call pattern.beforeMethod and pattern.afterMethod for inhereted static method with multiple aspect 2`] = `
Array [
  Object {
    "args": Array [],
    "context": [Function],
    "object": Object {
      "staticMethod": [Function],
    },
    "payload": "static method",
    "property": "staticMethod",
    "target": [Function],
  },
]
`;

exports[`aopForStatic method should call pattern.beforeMethod and pattern.afterMethod for own static method with multiple aspect 1`] = `
Array [
  Object {
    "args": Array [],
    "context": [Function],
    "object": Object {
      "staticMethod": [Function],
      "staticMethod2": [Function],
    },
    "property": "staticMethod2",
    "target": [Function],
  },
]
`;

exports[`aopForStatic method should call pattern.beforeMethod and pattern.afterMethod for own static method with multiple aspect 2`] = `
Array [
  Object {
    "args": Array [],
    "context": [Function],
    "object": Object {
      "staticMethod": [Function],
      "staticMethod2": [Function],
    },
    "payload": "static method 2",
    "property": "staticMethod2",
    "target": [Function],
  },
]
`;

exports[`aopForStatic method should call pattern.beforeMethod and pattern.afterMethod with multiple aspect 1`] = `
Array [
  Object {
    "args": Array [],
    "context": [Function],
    "object": Object {
      "staticMethod": [Function],
      "staticMethod2": [Function],
    },
    "property": "staticMethod",
    "target": [Function],
  },
]
`;

exports[`aopForStatic method should call pattern.beforeMethod and pattern.afterMethod with multiple aspect 2`] = `
Array [
  Object {
    "args": Array [],
    "context": [Function],
    "object": Object {
      "staticMethod": [Function],
      "staticMethod2": [Function],
    },
    "payload": "static method",
    "property": "staticMethod",
    "target": [Function],
  },
]
`;

exports[`aopForStatic method should call pattern.beforeSetter and pattern.afterSetter for inhereted static setter 2`] = `
Array [
  Object {
    "context": undefined,
    "object": Object {
      "staticMethod": [Function],
      "staticMethod2": [Function],
    },
    "payload": "static setter",
    "property": "staticSetter",
    "target": [Function],
  },
]
`;

exports[`aopForStatic method should call pattern.beforeSetter and pattern.afterSetter for inhereted static setter 3`] = `
Array [
  Object {
    "context": undefined,
    "object": Object {
      "staticMethod": [Function],
      "staticMethod2": [Function],
    },
    "payload": "static setter",
    "property": "staticSetter",
    "target": [Function],
  },
]
`;

exports[`aopForStatic method should not call pattern.afterMethod for static getter which return constructable function 1`] = `undefined`;

exports[`aopForStatic method should not call pattern.beforeGetter and pattern.afterGetter for not hooked property 2`] = `undefined`;

exports[`aopForStatic method should not call pattern.beforeGetter and pattern.afterGetter for not hooked property 3`] = `undefined`;
